pinc/upload_file.inc

```diff
@@ -0,0 +1,276 @@
+        . "<br>" . _("The rest of the file's name must consist of ASCII letters, digits, underscores, and/or hyphens. It must not begin with a hyphen and be no longer than 200 characters.");
+
+    $submit_blurb = "<p>" . sprintf(_("After you click the '%s' button, the browser will appear to be slow getting to the next page. This is because it is uploading the file."), $submit_label) . "</p>";
+
+    echo "<div style='border: 1px solid grey; margin-left: .5em; padding: .25em;'>\n";
+
+    echo "<form id='upload_form' method='POST' enctype='multipart/form-data'>\n";
+    echo "<input type='hidden' name='mode' value='upload'>";
+    echo $form_content;
+
+    echo "<p>$standard_blurb</p>";
+    // these are the form elements for non-resumable upload
+    echo "<div id='old_uploader'>
+    <p class='warning'>", sprintf(_('Maximum file size is %s.'), get_max_upload_size()), "</p>
+    <p>", get_big_upload_blurb(), "</p>",
+    _("File to upload"), ": <input type='file' id='old_browse' accept='.zip' name='uploaded_file'>
+    <input type='submit' id='old_submit' value='", attr_safe($submit_label), "'>",
+    "<p>", sprintf(_("After you click the '%s' button, the browser will appear to be slow getting to the next page. This is because it is uploading the file."), $submit_label), "</p>",
+    "</div>";
+
+    // if resumable, these are shown instead
+    echo "<div id='resumable_uploader' style='display: none;'>
+    <input type='hidden' name='resumable_filename' value=''>
+    <input type='hidden' name='resumable_identifier' value=''>
+    <p>", sprintf(_("Your browser supports uploading large files (up to %s) via javascript. If the upload fails,
+        or you navigate away from this page before it finishes, uploading the file again will pick up where it left off."), humanize_bytes(RESUMABLE_UPLOAD_SIZE)), "</p>
+    <p>", _("File to upload"), ": <span id='resumable_selected_file'></span> &nbsp;
+    <span id='resumable_browse' class='button'>", _("Choose File"), "</span>
+    <span id='resumable_submit' class='button'>", _("Upload"), "</span></p>
+    </div>";
+
+    echo "<p>", _("Upload progress"), ":&nbsp;<span id='upload_progress'></span></p>";
+
+    echo "</form>\n";
+    echo "</div>\n";
+}
+
+// This exception is thrown if problems are found in validate_uploaded_file()
+// or the other file-check functions below. It will contain text describing
+// the problem.
+class FileUploadException extends Exception {}
+
+class NoFileUploadedException extends Exception {}
+
+// This is called to process the posted information from the upload form
+// It returns a file_info with 'name' (the original file name)
+// and 'tmp_name' (the path to the file).
+// It checks for an empty file and for virus infection.
+// throws NoFileUploadedException if there is no file
+// throws a FileUploadException if there is a problem
+// if an exception is thrown any file will have been deleted.
+function validate_uploaded_file()
```

Comment:
One downside to this new implementation is that it doesn't give the user any feedback on upload status until everything is completed. In the current code, remote_file_manager pushes messages to the browser at each stage (file is uploaded, running AV check, etc). For the user's experience we should retain that more verbose output, at least for remote_file_manager.

Consider adding a `$verbose=FALSE` flag to this function. If `$verbose==TRUE`, output additional messages on the status, possibly with a helper function:
```
(in validate_uploaded_file)
        show_message($verbose, _("File successfully uploaded"));
        empty_check($file_size);
        show_message($verbose, _("Starting virus check"));
        virus_check($file_path);

function show_message($verbose, $message)
{
    if(!$verbose)
        return;
    echo "<p>$message</p>";
    flush();
}
```

To make this work, we need to disable the gzip encoding for the pages that use this. I would put that in `get_upload_args()`:
```
    // Disable gzip compression so we can flush the buffer after each step
    // in the process to give the user some progress details. Note that this
    // doesn't necessarily work for all browsers.
    apache_setenv('no-gzip', '1');
```

Reply:
I think it gives as much feedback as is useful. The only two things that take a significant amount of time are the upload itself and the virus check. It gives upload progress percentage messages and says "Upload complete. Running file checks, please wait..." while running the virus check. Any other messages would get immediately superseded.

Reply:
That's a good point, but it does presume that all uploads will be JS-based. That isn't an unreasonable expectation these days. It does also require that the browser leave the current page up while the virus scan is running. Chrome does, I'm not sure about the other browsers. This might also be impacted by the web server's gzip setting. It's on in TEST & PROD, but if it's off the server (which is the Apache default) posting to the next page may return a blank page while the virus scan runs.

All of that to say that there are a fair amount of unknowns around the user's experience in different scenarios.

Reply:
If it is uploading the old way it says "Working, please wait..." until everything is finished.
I don't know much about what is going on at lower levels but flushing would mean we were deliberately outputting invalid html in the sense that the end body and html tags were missing but browsers seem to handle it.

Reply:
> I don't know much about what is going on at lower levels but flushing would mean we were deliberately outputting invalid html

Less "invalid" and more "not valid yet", and this as always worked (for some definition of "always"). Think of a big 4MB HTML file - the browser will start rendering it to the user as soon as it has some content, refreshing as more content became available, well before the last body or html tag showed up.

Reply:
Ok, I don't know what would be best then. Let's see what srjfoo's testing reveals.

Reply:
I've been mulling this over, and I'm not sure what you mean @cpeel. I suspect that I just don't understand where it would be used where it might be trying to display a 4MB HTML file (or is that hypothetical and aimed at future-proofing?).

Reply:
Casey was explaining to me why my objection to outputting incomplete html pages did not stand up. The original issue was considering whether it would be possible/useful to show any more messages describing the progress of the operation.

Reply:
Okay, thanks. I think I understand now. As I said, my testing seemed to show successful or unsuccessful uploads depending on whether or not I was knowingly uploading a too-large zip. I do appreciate the verbose messages in Remote File Manager.

Reply:
Done something very like this. Had to change name to avoid conflict.